sdk/include/opentelemetry/sdk/common/http_client.h

```diff
@@ -0,0 +1,129 @@
+#include <string>
+#include <vector>
+#include "opentelemetry/nostd/function_ref.h"
+#include "opentelemetry/version.h"
+
+// Usage Example
+//
+// struct ReponseHandler: public HttpResponseHandler {
+//      void OnHttpResponse(HttpResponse& res) noexcept override
+//      {
+//           if (res.IsSuccess())
+//              .. process response
+//      }
+//  };
+//
+//   HttpSessionManager sessionManager; // implementer can provide singleton implementation for it
+//   auto session = sessionManager.createHttpSession("localhost", 8000);
+//   auto request = session->CreateRequest();
+//   request->AddHeader(..);
+//   ResponseHandler res_handler;
+//   session->SendRequest(res_handler);
+//   session->FinishSession() // optionally in the end.
+//   ...shutdown
+//    sessionManager.FinishAllSessions()
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace sdk
+{
+namespace common
+{
+namespace http
+{
+
+enum class HttpMethod
+{
+  HTTP_GET,
+  HTTP_POST,
+  HTTP_PUT,
+  HTTP_OPTIONS,
+  HTTP_HEAD,
+  HTTP_PATCH,
+  HTTP_DELETE
+};
+
+enum class HttpSessionState
+{
+  CREATED,  // session object is created
+  ONGOING,  // http(s) request is ongoing
+  FINISHED  // https(s) request is finished
+};
+
+using HttpHeader     = std::pair<std::string, std::string>;
+using HttpBody       = std::vector<uint8_t>;
+using HttpStatusCode = uint16_t;
+
+class HttpRequest
+{
+
+public:
+  virtual void setMethod(HttpMethod method) noexcept = 0;
+
+  virtual void setURI(const std::string &uri) noexcept = 0;
+
+  virtual void SetBody(HttpBody &body) noexcept = 0;
+
+  virtual void AddHeader(const HttpHeader &header) noexcept = 0;
+
+  virtual void ReplaceHeader(const HttpHeader &header) noexcept = 0;
+
+  virtual void SetTimeoutMs(std::chrono::milliseconds timeout_ms) noexcept = 0;
+};
+
+class HttpResponse
+{
+public:
+  virtual const HttpBody &GetBody() const noexcept = 0;
+
+  virtual bool GetNextHeader(nostd::function_ref<bool(HttpHeader)> callable) const noexcept = 0;
+
+  virtual bool GetNextHeader(const std::string &key,
+                             nostd::function_ref<bool(HttpHeader)> callable) const noexcept = 0;
+
+  virtual HttpStatusCode GetStatusCode() const noexcept = 0;
+
+  virtual bool IsSuccess() const noexcept = 0;
+
+  virtual const std::string &GetErrorMessage() const noexcept = 0;
+};
+
+class HttpResponseHandler
+{
+
+public:
+  virtual void OnHttpResponse(HttpResponse &) noexcept = 0;
+};
+
+class HttpSession
+{
+
+public:
+  virtual std::shared_ptr<HttpRequest> CreateRequest() noexcept = 0;
+
+  virtual void SendRequest(HttpResponseHandler &) noexcept = 0;
+
+  virtual HttpSessionState GetSessionState() const noexcept = 0;
+
+  virtual bool CancelSession() noexcept = 0;
+
+  virtual bool FinishSession() noexcept = 0;
+};
+
+class HttpSessionManager
+{
+
+public:
+  virtual std::shared_ptr<HttpSession> createHttpSession(std::string host,
+                                                         uint16_t port = 80) noexcept = 0;
```

Comment:
nit: do we need provide an override for HTTPS which is also very common?

Reply:
:+1: I actually think HTTPS is more important then HTTP. I know of many scenarios where only HTTPS is accepted.

Reply:
The plan was not to explicitly overload this method for HTTPS.  The `host` argument is supposed to (optionally) include http scheme along with server. So `https://localhost` defaults to HTTPS, and `http://localhost` or `localhost` defaults to HTTP. Probably, needs to be documented clearly. Let me know if you think otherwise.

Reply:
Also: does this method always return a session? Even if the host can't be resolved? Or is a a null check necessary (as apparently the method is `noexcept`)?

Reply:
Yes the `null` check would be necessary here. Also, it depends to some extent on the underlying http library been used. As not, all libraries provide api to create session initially - in that case `HttpSession` would just be returned as logical object instance without any established connection. And then actual connection would be made as part of `SendRequest()` method.